eip-5564.md

```diff
@@ -8,174 +8,170 @@ status: Draft
+  /// @dev The `entropy` is passed as `bytes memory` to support various keys formats and encryption
+  /// schemes. Consult the generator's documentation for details on how to format this parameter.
+  function generateStealthAddress(address registrant, bytes memory entropy)
+    external
+    view
+    returns (
+      address stealthAddress,
+      uint256 sharedSecret,
+      bytes32 sharedSecretHash,
+      bytes32 encryptedEntropy,
+      bytes memory ephemeralPubKey
+    );
+}
 
----
-
-
-#### For receiving (executed locally):
-
-for all `Announcement` events do:
+/// @notice Interface for sending tokens to a stealth address.
+interface IERC5564Messenger {
+  /// @dev Emitted when sending something to a stealth address.
+  /// @dev Metadata is an arbitrary field that the sender can use however they like, but the below
+  /// guidelines are recommended:
+  ///   - When sending ERC-20 tokens, the metadata SHOULD include the token address as the first 20
+  ///     bytes, and the amount being sent as the following 32 bytes.
+  ///   - When sending ERC-721 tokens, the metadata SHOULD include the token address as the first 20
+  ///     bytes, and the token ID being sent as the following 32 bytes.
+  event Announcement(
+    address indexed stealthRecipient,
+    bytes32 encryptedEntropy,
+    bytes ephemeralPubKey,
+    bytes metadata
+  );
+}
 
-if $pubtoaddr(P + (G * keccak(S * p)))$ == $stealthAddressRecipient$:
-```solidity
-bytes32 hashedSecret = keccak256(S);
-bytes sharedSecret = ecMul(hashedSecret, p);
-bytes pubkeyFromSharedSecret = ecMul(sharedSecret, G);
-bytes stealthPubkeyRecipient = ecAdd(pubkeyRecipient, pubkeyFromSharedSecret);
-address stealthAddressRecipient = address(stealthPubkeyRecipient);
-store_key_locally(p + keccak(S * p));
+/// @notice Sample implementation for generating stealth addresses for keys on the secp256k1 curve.
+function generateStealthAddress(address registrant, uint256 entropy, bytes memory ephemeralPrivKey)
+  external
+  view
+  returns (
+    address stealthAddress,
+    uint256 sharedSecret,
+    bytes32 sharedSecretHash,
+    bytes32 encryptedEntropy,
+    bytes memory ephemeralPubKey
+  )
+{
+  // Get user's public keys.
+  (bytes memory viewPubKey, bytes memory spendPubKey) =
+    IERC5564Registry(registry).stealthKeys(registrant, address(this));
+
+  // Generate shared secret from sender's private key and recipient's viewing key.
+  sharedSecret = ecMul(ephemeralPrivKey, viewPubKey);
+  sharedSecretHash = keccak256(sharedSecret);
+
+  // Encrypt the entropy using the shared secret.
+  encryptedEntropy = sharedSecret ^ entropy;
+
+  // Generate sender's public key from their ephemeral private key.
+  ephemeralPubKey = ecMul(ephemeralPrivKey, G);
+
+  // Compute stealth address from recipient's spending key and entropy.
+  stealthAddress = ecMul(spendPubKey, entropy);
+}
 ```
 
+Stealth addresses are computed using the algorithm below, assuming elliptic curves.
```

Comment:
This is the same algorithm Umbra currently uses (should double check it matches though, there may be typos), but definitely open to alternative algorithms if there's a reason to switch